packages/@aws-cdk/aws-route53/lib/vpc-endpoint-service-domain-name.ts

```diff
@@ -0,0 +1,198 @@
+// eslint-disable-next-line
+import { Construct as CoreConstruct } from '@aws-cdk/core';
+
+/**
+ * Properties to configure a VPC Endpoint Service domain name
+ */
+export interface VpcEndpointServiceDomainNameProps {
+
+  /**
+   * The VPC Endpoint Service to configure Private DNS for
+   */
+  readonly endpointService: IVpcEndpointService;
+
+  /**
+   * The domain name to use.
+   *
+   * This domain name must be verifiably owned by this account, or otherwise
+   * delegated to this account.
+   * @see https://docs.aws.amazon.com/vpc/latest/userguide/endpoint-services-dns-validation.html
+   */
+  readonly domainName: string;
+
+  /**
+   * The public hosted zone to use for the domain.
+   */
+  readonly publicHostedZone: IPublicHostedZone;
+}
+
+/**
+ * A Private DNS configuration for a VPC endpoint service.
+ */
+export class VpcEndpointServiceDomainName extends CoreConstruct {
+
+  // Track all domain names created, so someone doesn't accidentally associate two domains with a single service
+  private static readonly endpointServices: IVpcEndpointService[] = [];
+
+  // The way this class works is by using three custom resources and a TxtRecord in conjunction
+  // The first custom resource tells the VPC endpoint service to use the given DNS name
+  // The VPC endpoint service will then say:
+  // "ok, create a TXT record using these two values to prove you own the domain"
+  // The second custom resource retrieves these two values from the service
+  // The TxtRecord is created from these two values
+  // The third custom resource tells the VPC Endpoint Service to verify the domain ownership
+  constructor(scope: Construct, id: string, props: VpcEndpointServiceDomainNameProps) {
+    super(scope, id);
+
+    // Make sure a user doesn't accidentally add multiple domains
+    this.validateProps(props);
+    VpcEndpointServiceDomainName.endpointServices.push(props.endpointService);
+
+    const serviceId = props.endpointService.vpcEndpointServiceId;
+    const privateDnsName = props.domainName;
+
+    // Enable Private DNS on the endpoint service and retrieve the AWS-generated configuration
+    const privateDnsConfiguration = this.getPrivateDnsConfiguration(id, serviceId, privateDnsName);
+
+    // Tell AWS to verify that this account owns the domain attached to the service
+    this.verifyPrivateDnsConfiguration(privateDnsConfiguration, props.publicHostedZone);
+
+    // Finally, don't do any of the above before the endpoint service is created
+    this.node.addDependency(props.endpointService);
+  }
+
+  private validateProps(props: VpcEndpointServiceDomainNameProps): void {
+    if (VpcEndpointServiceDomainName.endpointServices.includes(props.endpointService)) {
+      throw new Error(
+        'Cannot create a VpcEndpointServiceDomainName for service ' +
+        props.endpointService.node.uniqueId +
+        ', another VpcEndpointServiceDomainName is already associated with it');
```

Comment:
prefer string interpolation to concatenation - perhaps also provide the id that's already associated as well?

```suggestion
        `Cannot create a VpcEndpointServiceDomainName for service '${props.endpointService.node.uniqueId}', another VpcEndpointServiceDomainName is already associated with it`);
```

Reply:
@shivlaks that's a good idea, is it possible to get the "id" that the user provided as a constructor to the other endpoint service construct? Would the unique ID be that?

Reply:
Changed the error to mention the other endpoint domain name in the error message.